engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/IfThenElseNode.java

```diff
@@ -0,0 +1,154 @@
+    this.cond = Objects.requireNonNull(cond);
+    this.trueBranch = trueBranch;
+    this.falseBranch = falseBranch;
+    this.with = IfThenElseNodeFactory.WithConditionNodeGen.create();
+  }
+
+  public static IfThenElseNode build(ExpressionNode c, ExpressionNode t, ExpressionNode fOrNull) {
+    return new IfThenElseNode(c, t, fOrNull);
+  }
+
+  @Override
+  public Object executeGeneric(VirtualFrame frame) {
+    var condValue = cond.executeGeneric(frame);
+    var result = with.executeIf(frame, condValue, trueBranch, falseBranch);
+    return result;
+  }
+
+  abstract static class WithCondition extends Node {
+    WithCondition() {}
+
+    abstract Object executeIf(
+        VirtualFrame frame, Object cond, ExpressionNode trueBranch, ExpressionNode falseBranch);
+
+    @Specialization
+    final Object doBoolean(
+        VirtualFrame frame,
+        boolean cond,
+        ExpressionNode trueBranch,
+        ExpressionNode falseBranch,
+        @Shared("profile") @Cached InlinedCountingConditionProfile profile) {
+      if (cond) {
+        profile.wasTrue(this);
+        return trueBranch.executeGeneric(frame);
+      } else {
+        profile.wasFalse(this);
+        if (falseBranch == null) {
+          var ctx = EnsoContext.get(this);
+          return ctx.getBuiltins().nothing();
+        } else {
+          return falseBranch.executeGeneric(frame);
+        }
+      }
+    }
+
+    static boolean notEnsoObject(Object o) {
+      return !(o instanceof EnsoObject);
+    }
+
+    @Specialization(
+        limit = "3",
+        guards = {"notEnsoObject(foreignObj)", "iop.isBoolean(foreignObj)"})
+    final Object doTruffleBoolean(
+        VirtualFrame frame,
+        TruffleObject foreignObj,
+        ExpressionNode trueBranch,
+        ExpressionNode falseBranch,
+        @CachedLibrary("foreignObj") InteropLibrary iop,
+        @Shared("profile") @Cached InlinedCountingConditionProfile profile) {
+      try {
+        var cond = iop.asBoolean(foreignObj);
+        return doBoolean(frame, cond, trueBranch, falseBranch, profile);
+      } catch (UnsupportedMessageException ex) {
+        var ctx = EnsoContext.get(this);
+        throw ctx.raiseAssertionPanic(this, "Expecting boolean", ex);
+      }
+    }
+
+    @Specialization(
+        limit = "3",
+        guards = {"cond != null", "warnings.hasWarnings(cond)"})
+    final Object doWarning(
+        VirtualFrame frame,
+        Object cond,
+        ExpressionNode trueBranch,
+        ExpressionNode falseBranch,
+        @CachedLibrary(value = "cond") WarningsLibrary warnings,
+        @Cached AppendWarningNode appendWarningNode,
+        @Cached WithCondition delegate) {
+      try {
+        var ws = warnings.getWarnings(cond, false);
+        var without = warnings.removeWarnings(cond);
+
+        var result = delegate.executeIf(frame, without, trueBranch, falseBranch);
+        return appendWarningNode.executeAppend(null, result, ws);
```

Comment:
Why not pass the `frame` along? Why `null`?

Reply:
Thanks. Updated in 933d74ca7e